File: app/src/main/java/com/druger/aboutwork/utils/rx/RxSearch.java
package com.druger.aboutwork.utils.rx;

import androidx.annotation.NonNull;
import androidx.appcompat.widget.SearchView;

import io.reactivex.Observable;
import io.reactivex.subjects.BehaviorSubject;

/**
 * Created by druger on 05.07.2017.
 */

public class RxSearch {

    private RxSearch() {
    }

    public static Observable<String> fromSearchView(@NonNull SearchView searchView) {
        final BehaviorSubject<String> subject = BehaviorSubject.create();

        searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String query) {
                subject.onComplete();
                return true;
            }

            @Override
            public boolean onQueryTextChange(String newText) {
                if (!newText.isEmpty()) {
                    subject.onNext(newText);
                }
                return true;
            }
        });
        return subject;
    }
}

File: app/src/main/java/com/druger/aboutwork/utils/rx/RxSearch.kt
package com.druger.aboutwork.utils.rx

import androidx.appcompat.widget.SearchView

import io.reactivex.Observable
import io.reactivex.subjects.BehaviorSubject

/**
 * Created by druger on 05.07.2017.
 */

object RxSearch {

    fun fromSearchView(searchView: SearchView): Observable<String> {
        val subject = BehaviorSubject.create<String>()

        searchView.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
            override fun onQueryTextSubmit(query: String): Boolean {
                subject.onComplete()
                return true
            }

            override fun onQueryTextChange(newText: String): Boolean {
                if (newText.isNotEmpty()) {
                    subject.onNext(newText)
                }
                return true
            }
        })
        return subject
    }
}
